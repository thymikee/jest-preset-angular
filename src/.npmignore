# Tests
__tests__

# sources are inlined
zone-patch
AngularNoNgAttributesSnapshotSerializer.js
AngularSnapshotSerializer.js
HTMLCommentSerializer.js
*.ts
*.tgz
