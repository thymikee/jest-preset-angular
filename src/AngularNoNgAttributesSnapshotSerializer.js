'use strict';

const jestDOMElementSerializer = require('pretty-format').plugins.DOMElement;

const attributesToRemovePatterns = ['ng-reflect', '_nghost', '_ngcontent', 'ng-version'];
const attributesToClean = {
  class: [/^(?:mat|cdk|ng).*-\w*\d+-\d+$/, /^ng-star-inserted$/], // e.g. "ng-tns-c25-1" or "ng-star-inserted", literally
  id: [/^(?:mat|cdk|ng).*-\d+$/], // e.g. "mat-input-4", "cdk-step-content-0-0"
  for: [/^(?:mat|cdk|ng).*-\d+$/], // e.g. "mat-form-field-label-9"
  'aria-owns': [/^(?:mat|cdk|ng).*-\d+$/], // e.g. "mat-input-4"
  'aria-labelledby': [/^(?:mat|cdk|ng).*-\d+$/], // e.g. "mat-input-4", "cdk-step-label-0-0"
  'aria-controls': [/^(?:mat|cdk|ng).*-\d+$/], // e.g. "cdk-step-content-2-0"
};

const hasAttributesToRemove = (attribute) =>
  attributesToRemovePatterns.some((removePattern) => attribute.name.startsWith(removePattern));
const hasAttributesToClean = (attribute) =>
  Object.keys(attributesToClean).some((removePatternKey) => attribute.name === removePatternKey);

const serialize = (node, ...rest) => {
  const nodeCopy = node.cloneNode(true);
  // Remove angular-specific attributes
  Object.values(nodeCopy.attributes)
    .filter(hasAttributesToRemove)
    .forEach((attribute) => nodeCopy.attributes.removeNamedItem(attribute.name));
  // Remove angular auto-added classes
  Object.values(nodeCopy.attributes)
    .filter(hasAttributesToClean)
    .forEach((attribute) => {
      attribute.value = attribute.value
        .split(' ')
        .filter(
          (attrValue) =>
            !attributesToClean[attribute.name].some((attributeCleanRegex) => attributeCleanRegex.test(attrValue))
        )
        .join(' ');
      if (attribute.value === '') {
        nodeCopy.attributes.removeNamedItem(attribute.name);
      } else {
        nodeCopy.attributes.setNamedItem(attribute);
      }
    });

  return jestDOMElementSerializer.serialize(nodeCopy, ...rest);
};

const serializeTestFn = (val) =>
  val.attributes !== undefined &&
  Object.values(val.attributes).some(
    (attribute) => hasAttributesToRemove(attribute) || hasAttributesToClean(attribute)
  );
const test = (val) => jestDOMElementSerializer.test(val) && serializeTestFn(val);

module.exports = {
  serialize,
  test,
};
