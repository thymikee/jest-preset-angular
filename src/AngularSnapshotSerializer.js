'use strict';

const printAttributes = (val, attributes, print, indent, colors, opts) => {
  return attributes
    .sort()
    .map(attribute => {
      return (
        opts.spacing +
        indent(colors.prop.open + attribute + colors.prop.close + '=') +
        colors.value.open +
        (val.componentInstance[attribute] &&
          val.componentInstance[attribute].constructor
          ? '{[Function ' +
              val.componentInstance[attribute].constructor.name +
              ']}'
          : `"${val.componentInstance[attribute]}"`) +
        colors.value.close
      );
    })
    .join('');
};

const print = (val, print, indent, opts, colors) => {
  let componentAttrs = '';

  const componentName = val.componentRef._elDef.element.name;
  const nodes = (val.componentRef._view.nodes || [])
    .filter(node => node && node.hasOwnProperty('renderElement'))
    .map(node => Array.from(node.renderElement.childNodes).map(print).join(''))
    .join(opts.edgeSpacing);

  const attributes = Object.keys(val.componentInstance);

  if (attributes.length) {
    componentAttrs += printAttributes(
      val,
      attributes,
      print,
      indent,
      colors,
      opts
    );
  }

  return (
    '<' +
    componentName +
    componentAttrs +
    (componentAttrs.length ? '\n' : '') +
    '>\n' +
    indent(nodes) +
    '\n</' +
    componentName +
    '>'
  );
};

const test = val =>
  val !== undefined &&
  val !== null &&
  typeof val === 'object' &&
  Object.prototype.hasOwnProperty.call(val, 'componentRef');

module.exports = {
  print: print,
  test: test
};
