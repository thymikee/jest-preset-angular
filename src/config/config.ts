import type { TsJestTransformerOptions } from 'ts-jest';

export type NgJestTransformerOptions = TsJestTransformerOptions & {
    processWithEsbuild?: string[];
};
