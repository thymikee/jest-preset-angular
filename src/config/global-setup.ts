import { runNgccJestProcessor } from '../utils/ngcc-jest-processor';

export const globalSetup = async () => {
  const ngJestConfig = globalThis.ngJest;
  const tsconfig = ngJestConfig?.tsconfig;
  if (!ngJestConfig?.skipNgcc) {
    runNgccJestProcessor(tsconfig);
  }
};
