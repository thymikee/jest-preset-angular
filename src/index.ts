import { NgJestTransformer } from './ng-jest-transformer';

export default {
  createTransformer: (): NgJestTransformer => new NgJestTransformer(),
};
