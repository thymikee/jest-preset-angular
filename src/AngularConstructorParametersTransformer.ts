import type {
  TransformerFactory,
  SourceFile,
  TransformationContext,
  Program,
} from "typescript";
import { ConfigSet } from "ts-jest/dist/config/config-set";

function tryLoadV8V9Transformer(
  program: Program
): TransformerFactory<SourceFile> | undefined {
  try {
    const {
      downlevelConstructorParameters,
    } = require("@ngtools/webpack/src/transformers/ctor-parameters");

    return downlevelConstructorParameters(() => program.getTypeChecker());
  } catch (err) {
    return undefined;
  }
}

function tryLoadV10Transformer(
  program: Program
): TransformerFactory<SourceFile> | undefined {
  try {
    const {
      constructorParametersDownlevelTransform,
    } = require("@angular/compiler-cli/src/tooling");

    return (context: TransformationContext) => {
      const compilerCliFactory = constructorParametersDownlevelTransform(
        program
      );
      return compilerCliFactory(context);
    };
  } catch (err) {
    return undefined;
  }
}

/**
 * The factory of hoisting transformer factory
 * @internal
 */
export function factory(configSet: ConfigSet): TransformerFactory<SourceFile> {
  const { program } = configSet.tsCompiler;
  if (!program) {
    throw new Error(
      "TypeScript Compiler module is not provided! Please do not use the `AngularConstructorParametersTransformer` with `isolatedModules: true`"
    );
  }

  // Depending on the used Angular version, we have to load the constructor parameters transformer from different locations

  // Angular v8/v9 downlevel constructor transformer
  let transformerFactory = tryLoadV8V9Transformer(program);

  // Angular v10+ downlevel constructor transformer
  if (transformerFactory === undefined) {
    transformerFactory = tryLoadV10Transformer(program);
  }

  if (transformerFactory === undefined) {
    throw new Error(
      "Could not load Constructor Parameters Downlevel Transformer!"
    );
  }

  return transformerFactory;
}
