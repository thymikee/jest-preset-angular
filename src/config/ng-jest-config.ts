import type { ParsedConfiguration } from '@angular/compiler-cli';
import { formatDiagnostics, readConfiguration } from '@angular/compiler-cli';
import type { Config } from '@jest/types';
import { ConfigSet } from 'ts-jest/dist/config/config-set';
import { ParsedCommandLine, ModuleKind, ScriptTarget } from 'typescript';

export class NgJestConfig extends ConfigSet {
  private parsedNgConfig!: ParsedConfiguration;

  constructor(private jestConfig: Config.ProjectConfig) {
    super(jestConfig);
    this.setupOptions(jestConfig);
  }

  /**
   * Override `ts-jest` parsedTsConfig so that it always returns Angular compiler config
   */
  get parsedTsConfig(): ParsedConfiguration {
    return this.parsedNgConfig;
  }

  private setupOptions(jestConfig: Config.ProjectConfig): void {
    const { globals } = jestConfig;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const extraTsconfig: string | ParsedCommandLine | undefined =
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      globals && globals['ts-jest'] ? (globals as Record<string, any>)['ts-jest'].tsconfig : undefined;
    /**
     * By default, readConfiguration will pick up `tsconfig.json` from current directory. Because we are getting config
     * via `ts-jest`, we need to manually get `tsconfig` from `ts-jest` options from `jest` config.
     */
    if (typeof extraTsconfig === 'string') {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error `ts-jest` doesn't expose typing for resolvePath
      this.parsedNgConfig = readConfiguration(super.resolvePath(extraTsconfig));
    } else {
      this.parsedNgConfig = readConfiguration(this.jestConfig.cwd ?? process.cwd());
      if (extraTsconfig) {
        this.parsedNgConfig = {
          ...this.parsedNgConfig,
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          options: {
            ...this.parsedNgConfig.options,
            ...extraTsconfig,
          } as Record<string, any>, // eslint-disable-line @typescript-eslint/no-explicit-any
        };
      }
    }
    if (this.parsedNgConfig.errors?.length) {
      throw new Error(formatDiagnostics(this.parsedNgConfig.errors));
    }
    this.parsedNgConfig = {
      ...this.parsedNgConfig,
      options: {
        ...this.parsedNgConfig.options,
        module: ModuleKind.CommonJS,
        target: ScriptTarget.ES5,
      },
    };
  }
}
