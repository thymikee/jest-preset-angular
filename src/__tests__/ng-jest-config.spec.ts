import { join } from 'path';

import { normalizeSeparators } from '@angular/compiler-cli/src/ngtsc/util/src/path';

import { NgJestConfig } from '../config/ng-jest-config';

describe('NgJestConfig', () => {
  const specifiedTsCfgPath = join(process.cwd(), 'tsconfig.spec.json');
  const defaultTsCfgPath = join(process.cwd(), 'tsconfig.json');

  describe('_resolveTsConfig', () => {
    test('should return config including Angular compiler config with tsconfig as a string from ts-jest option', () => {
      const ngJestConfig = new NgJestConfig({
        testMatch: [],
        testRegex: [],
        extensionsToTreatAsEsm: [],
        globals: {
          'ts-jest': {
            tsconfig: specifiedTsCfgPath,
          },
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any);
      const config = ngJestConfig.parsedTsConfig;

      delete config.options.basePath;
      delete config.options.baseUrl;
      expect(config.options.configFilePath).toEqual(normalizeSeparators(specifiedTsCfgPath));
      delete config.options.configFilePath;
      delete config.options.genDir;
      expect(config.options).toMatchSnapshot();
    });

    test.each([
      {},
      {
        skipLibCheck: false,
      },
    ])(
      'should return config including Angular compiler config with tsconfig as an object from ts-jest option',
      (tsconfig) => {
        const ngJestConfig = new NgJestConfig({
          cwd: '.',
          testMatch: [],
          testRegex: [],
          extensionsToTreatAsEsm: [],
          globals: {
            'ts-jest': {
              tsconfig,
            },
          },
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } as any);
        const config = ngJestConfig.parsedTsConfig;

        delete config.options.basePath;
        delete config.options.baseUrl;
        expect(config.options.configFilePath).toEqual(normalizeSeparators(defaultTsCfgPath));
        delete config.options.configFilePath;
        delete config.options.genDir;
        expect(config.options).toMatchSnapshot();
      },
    );

    test('should return config including Angular compiler config without tsconfig defined in ts-jest option', () => {
      const ngJestConfig = new NgJestConfig({
        testMatch: [],
        testRegex: [],
        extensionsToTreatAsEsm: [],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any);
      const config = ngJestConfig.parsedTsConfig;

      delete config.options.basePath;
      delete config.options.baseUrl;
      expect(config.options.configFilePath).toEqual(normalizeSeparators(defaultTsCfgPath));
      delete config.options.configFilePath;
      delete config.options.genDir;
      expect(config.options).toMatchSnapshot();
    });
  });
});
