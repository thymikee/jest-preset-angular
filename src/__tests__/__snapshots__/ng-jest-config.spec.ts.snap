// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgJestConfig readTsConfig should return config including Angular compiler config with tsconfig as a string from ts-jest option 1`] = `
Object {
  "enableIvy": false,
  "fullTemplateTypeCheck": true,
  "module": 1,
  "strictInjectionParameters": true,
  "target": 1,
}
`;

exports[`NgJestConfig readTsConfig should return config including Angular compiler config with tsconfig as an object from ts-jest option 1`] = `
Object {
  "enableIvy": true,
  "fullTemplateTypeCheck": true,
  "module": 1,
  "resolveJsonModule": true,
  "strictInjectionParameters": true,
  "target": 1,
}
`;

exports[`NgJestConfig readTsConfig should return config including Angular compiler config without tsconfig defined in ts-jest option 1`] = `
Object {
  "enableIvy": true,
  "esModuleInterop": true,
  "forceConsistentCasingInFileNames": true,
  "importHelpers": true,
  "importsNotUsedAsValues": 2,
  "lib": Array [
    "lib.esnext.d.ts",
    "lib.dom.d.ts",
  ],
  "module": 1,
  "noFallthroughCasesInSwitch": true,
  "noImplicitReturns": true,
  "noUnusedLocals": true,
  "noUnusedParameters": true,
  "skipLibCheck": true,
  "strict": true,
  "target": 1,
  "types": Array [
    "node",
    "jest",
  ],
}
`;
